### PR TITLE
Create managed resources for controlplane

### LINKED ISSUE #111: Ordering of controlplane creation and kube-apiserver deployment steps
We encountered an ordering issue when trying to add controlplane creation to the Gardener flow:

* Creation of controlplane must happen *before* deployment of kube-apiserver due to some providers mounting the `cloud-provider-config` configmap as a volume (via webhooks).
* At the same time it must happen *after* deployment of kube-apiserver due to the logic in controlplane controllers that deploys controlplane-related provider-specific resources to the Shoot.

We should resolve this issue in the following way:

* Initially, we should disable deploying to the Shoot by controlplane controllers. For the time being, Gardener should continue to deploy the corresponding provider-specific resources to the Shoot as before (as kube-addon-manager addons).
* As a second step, controlplane controllers should write the corresponding manifests as "managed resources" (a new CRD) to the Shoot namespace. 
* As a third step, we will implement `gardener-resource-manager`, a new controller that supercedes kube-addon-manager and takes care of actually deploying these resources to the Shoot. At that point, we should also remove the corresponding logic from Gardener.

### PR DESCRIPTION
**Pull request by stoyanr**
**What this PR does / why we need it**:
Enhances controlplane controllers to create managed resources for the shoot cluster.

**Which issue(s) this PR fixes**:
Fixes #111 

**Special notes for your reviewer**:
This is now properly tested and all dependencies are in, I consider it ready for review and merge.

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

File: pkg/controller/controlplane/genericactuator/actuator.go [hunk 1]
```diff
@@ -200,27 +206,47 @@ func (a *actuator) Reconcile(
...
+		return false, errors.Wrapf(err, "could not render control plane shoot chart for controlplane '%s'", util.ObjectName(cp))
+	}
+
+	// Create or update secret containing the rendered control plane shoot chart
+	a.logger.Info("Creating secret", "controlplane", util.ObjectName(cp), "name", resourceName)
+	if err := resourcemanager.NewSecret(a.client).
+		WithNamespacedName(types.NamespacedName{Namespace: cp.Namespace, Name: resourceName}).
+		WithKeyValues(map[string][]byte{name: data}).
+		Reconcile(ctx); err != nil {
+		return false, errors.Wrapf(err, "could not create or update secret '%s/%s' for controlplane '%s'", cp.Namespace, resourceName, util.ObjectName(cp))
```
Comment (rfranzke) [inline pkg/controller/controlplane/genericactuator/actuator.go]:
```suggestion
		return false, errors.Wrapf(err, "could not create or update secret '%s/%s' of managed resource for controlplane '%s'", cp.Namespace, resourceName, util.ObjectName(cp))
```

File: pkg/controller/controlplane/genericactuator/actuator.go [hunk 2]
```diff
@@ -200,27 +206,47 @@ func (a *actuator) Reconcile(
...
+	// Render control plane shoot chart
+	a.logger.Info("Rendering control plane shoot chart", "controlplane", util.ObjectName(cp), "values", values)
+	version := cluster.Shoot.Spec.Kubernetes.Version
+	name, data, err := a.controlPlaneShootChart.Render(chartRenderer, metav1.NamespaceSystem, a.imageVector, version, version, values)
+	if err != nil {
+		return false, errors.Wrapf(err, "could not render control plane shoot chart for controlplane '%s'", util.ObjectName(cp))
+	}
+
+	// Create or update secret containing the rendered control plane shoot chart
+	a.logger.Info("Creating secret", "controlplane", util.ObjectName(cp), "name", resourceName)
```
Comment (rfranzke) [inline pkg/controller/controlplane/genericactuator/actuator.go]:
```suggestion
	a.logger.Info("Creating secret for managed resource containing shoot chart", "controlplane", util.ObjectName(cp), "name", resourceName)
```

File: pkg/controller/controlplane/genericactuator/actuator.go [hunk 3]
```diff
@@ -200,27 +206,47 @@ func (a *actuator) Reconcile(
...
+	if err := resourcemanager.NewSecret(a.client).
+		WithNamespacedName(types.NamespacedName{Namespace: cp.Namespace, Name: resourceName}).
+		WithKeyValues(map[string][]byte{name: data}).
+		Reconcile(ctx); err != nil {
+		return false, errors.Wrapf(err, "could not create or update secret '%s/%s' for controlplane '%s'", cp.Namespace, resourceName, util.ObjectName(cp))
+	}
+
+	// Create or update managed resource referencing the previously created secret
+	a.logger.Info("Creating managed resource", "controlplane", util.ObjectName(cp), "name", resourceName)
```
Comment (rfranzke) [inline pkg/controller/controlplane/genericactuator/actuator.go]:
```suggestion
	a.logger.Info("Creating managed resource for shoot chart", "controlplane", util.ObjectName(cp), "name", resourceName)
```

File: pkg/controller/controlplane/genericactuator/actuator.go [hunk 4]
```diff
@@ -200,27 +206,47 @@ func (a *actuator) Reconcile(
...
+	if err := resourcemanager.NewManagedResource(a.client).
+		WithNamespacedName(types.NamespacedName{Namespace: cp.Namespace, Name: resourceName}).
+		WithSecretRef(resourceName).
+		Reconcile(ctx); err != nil {
+		return false, errors.Wrapf(err, "could not create or update managed resource '%s/%s' for controlplane '%s'", cp.Namespace, resourceName, util.ObjectName(cp))
```
Comment (rfranzke) [inline pkg/controller/controlplane/genericactuator/actuator.go]:
```suggestion
		return false, errors.Wrapf(err, "could not create or update managed resource '%s/%s' containing shoot chart for controlplane '%s'", cp.Namespace, resourceName, util.ObjectName(cp))
```

File: pkg/controller/controlplane/genericactuator/actuator.go [hunk 5]
```diff
@@ -232,17 +258,21 @@ func (a *actuator) Delete(
...
+	// Delete the managed resource
+	a.logger.Info("Deleting managed resource", "controlplane", util.ObjectName(cp), "name", resourceName)
+	if err := resourcemanager.NewManagedResource(a.client).
+		WithNamespacedName(types.NamespacedName{Namespace: cp.Namespace, Name: resourceName}).
+		Delete(ctx); err != nil {
+		return errors.Wrapf(err, "could not delete managed resource '%s/%s' for controlplane '%s'", cp.Namespace, resourceName, util.ObjectName(cp))
+	}
+
+	// Delete the secret referenced by the managed resource
+	a.logger.Info("Deleting secret", "controlplane", util.ObjectName(cp), "name", resourceName)
```
Comment (rfranzke) [inline pkg/controller/controlplane/genericactuator/actuator.go]:
```suggestion
	a.logger.Info("Deleting secret of managed resource", "controlplane", util.ObjectName(cp), "name", resourceName)
```

File: pkg/util/shoot_clients.go [hunk 6]
```diff
@@ -102,3 +107,27 @@ func NewClientsForShoot(ctx context.Context, c client.Client, namespace string,
...
+// NewChartRendererForShoot creates a new chartrenderer.Interface for the shoot cluster.
+func NewChartRendererForShoot(version string) (chartrenderer.Interface, error) {
+	v, err := VersionInfo(version)
+	if err != nil {
+		return nil, err
+	}
+	return chartrenderer.New(engine.New(), &chartutil.Capabilities{KubeVersion: v}), nil
+}
+
+var versionRegex = regexp.MustCompile(`^(\d+\.)?(\d+\.)?(\d+)$`)
```
Comment (rfranzke) [inline pkg/util/shoot_clients.go]: Should we rather use [Masterminds/semver](https://github.com/gardener/gardener/blob/master/Gopkg.toml#L119-L121) like in gardener?

**Comment by rfranzke** [comment]
gardener/gardener#1130 is merged and cherry-picked to the `gardener-extensions-fixes` branch, you can revendor to commit `f544094d821f4316f3840c142e924d35e9f08c7a` now.

**Comment by zanetworker** [comment]
@stoyanr I have moved the whole library to the https://github.com/gardener/gardener-resource-manager/tree/master/pkg/manager  since we might want to use it in Gardener as well. Please re-vendor.

**Comment by rfranzke** [comment]
gardener/gardener#1114 is merged

**Comment by stoyanr** [comment]
> @stoyanr I have moved the whole library to the https://github.com/gardener/gardener-resource-manager/tree/master/pkg/manager  since we might want to use it in Gardener as well. Please re-vendor.

@zanetworker Impossible, since `gardener-resource-manager` depends on Kubernetes 1.14 / `controller-runtime` / 0.2.0-beta.2, it's currently not possible for `gardener-extensions` to depend on that, and there are incompatible changes. Your library is currently unusable. Would you consider moving it back, or should I just go on without it?

File: pkg/controller/controlplane/genericactuator/actuator.go (hunk 1 above)
Comment (stoyanr) [reply pkg/controller/controlplane/genericactuator/actuator.go]: Done

File: pkg/controller/controlplane/genericactuator/actuator.go (hunk 2 above)
Comment (stoyanr) [reply pkg/controller/controlplane/genericactuator/actuator.go]: Done

File: pkg/controller/controlplane/genericactuator/actuator.go (hunk 3 above)
Comment (stoyanr) [reply pkg/controller/controlplane/genericactuator/actuator.go]: Done

File: pkg/controller/controlplane/genericactuator/actuator.go (hunk 4 above)
Comment (stoyanr) [reply pkg/controller/controlplane/genericactuator/actuator.go]: Done

File: pkg/controller/controlplane/genericactuator/actuator.go (hunk 5 above)
Comment (stoyanr) [reply pkg/controller/controlplane/genericactuator/actuator.go]: Done

File: pkg/util/shoot_clients.go (hunk 6 above)
Comment (stoyanr) [reply pkg/util/shoot_clients.go]: Good idea. In fact, `Mastermind/semver` was already used by `VersionMajorMinor`, so I refactored `VersionInfo` accordingly, moved it to `shoot.go`, and added a test.

**Comment by rfranzke** [comment]
... It's really cumbersome. I suggest not to move it to `Gardener` but simply proceed without it for now (directly define the `Secret`s and `ManagedResource`s and applying them).

**Comment by zanetworker** [comment]
@stoyanr you can try this commit hash https://github.com/gardener/gardener-resource-manager/blob/e0f770704e8c97f8a4a5003f14d88a1ae55253a1/Gopkg.toml 

It uses an older version of controller-runtime and has the library in it. This way you don't have to change your implementation till the issue is fixed, you just need to re-vendor later.

**Comment by stoyanr** [comment]
> @stoyanr you can try this commit hash https://github.com/gardener/gardener-resource-manager/blob/e0f770704e8c97f8a4a5003f14d88a1ae55253a1/Gopkg.toml
> 
> It uses an older version of controller-runtime and has the library in it. This way you don't have to change your implementation till the issue is fixed, you just need to re-vendor later.

Ok, done.

**Pull request merged**